Automattic/jetpack
# Widgets: Adds Simple Payment Buttons as Widgets

Original Feature Branch: https://github.com/Automattic/jetpack/pull/9577

Since this is a Feature branch, it fixes a lot of things: 
- https://github.com/Automattic/wp-calypso/issues/24617 (Simple Payments: Widget - Epic)
- https://github.com/Automattic/wp-calypso/issues/24614 (Customizer: Add Existing Simple Payment Button as Widget)
- https://github.com/Automattic/wp-calypso/issues/24676 (Customizer: Create a Simple Payments Button from the Customizer Widget UI)
- https://github.com/Automattic/wp-calypso/issues/24693 (Customizer: Edit existing Simple Payment Button on the Customizer UI)
- https://github.com/Automattic/wp-calypso/issues/24694 (Customizer: remove Simple Payments Button)
- https://github.com/Automattic/wp-calypso/issues/25342 (Customizer: add defaults to the Add New Simple Payment Button form)
- https://github.com/Automattic/wp-calypso/issues/25328 (Customizer: Add validations to the Create New/Edit Simple Payment Button form.)

Depends on:

- ~Simple Payments: adds plan support check for Jetpack and WordPress.com (#9650 and D14005-code)~

#### Changes proposed in this Pull Request:

Adds support on the Customizer to add Simple Payment Buttons as Sidebar Widgets.

| Widget List | Widget Customizer Form & Preview |
| - | - |
| ![screen shot 2018-05-18 at 12 14 37](https://user-images.githubusercontent.com/233601/40242922-5af9c0ac-5a95-11e8-9929-945e2c4e7e21.png) | ![screen shot 2018-05-18 at 12 15 34](https://user-images.githubusercontent.com/233601/40242925-5fb172b6-5a95-11e8-889b-4b5763f08170.png) |

Also, the user can manage their Simple Payment Buttons from the Customizer with a collapsible form

| Cosed | Open |
| - | - |
| ![screen shot 2018-06-06 at 15 10 32](https://user-images.githubusercontent.com/233601/41056991-3fa8a262-699c-11e8-89e4-ab28ce16e5e2.png) | ![screen shot 2018-06-06 at 15 11 15](https://user-images.githubusercontent.com/233601/41057003-466ed74c-699c-11e8-8012-2e934a2cc4e2.png) |

| Add New | Edit | Delete |
| - | - | - |
| ![my movie 1 mp4](https://user-images.githubusercontent.com/233601/41058102-4bb41be2-699f-11e8-8be6-b034a049f479.gif) | ![edit](https://user-images.githubusercontent.com/233601/41335035-e00442b8-6ebd-11e8-8d7f-8b79cf3f27d9.gif) | ![delete](https://user-images.githubusercontent.com/233601/41334902-7129293a-6ebd-11e8-8b4c-e924d6937c1d.gif) |

The form uses the same defaults and validations used on Calypso

| Post/Page Form | Customizer Form |
| - | - |
| ![screen shot 2018-06-07 at 01 28 33](https://user-images.githubusercontent.com/233601/41078364-b8162cb4-69f2-11e8-866d-48752b02604e.png) | ![screen shot 2018-06-18 at 17 56 23](https://user-images.githubusercontent.com/233601/41561760-2616c912-7321-11e8-8134-4bcc236af98d.png) |

#### Testing instructions:

Pre-requisites:
- apply this branch on a Jetpack site with an active Business Plan subscription and a Theme with at least one widget area
- to create one or more Payment Buttons on the Post/Page editor

then you'll need to:

- open the Customizer on your sandboxed Business Site
- go to Widgets and select a Widget Area
- click on Add Widget
- search on the widget panel for Simple Payment
- select the Simple Payment widget

The Simple Payment Customizer should list the available Payment Buttons, and the Customizer Preview should show the item selected by default. All changes should update live on the preview window and the site should only be updated upon publishing.

To **create a new product**, you'll need to:

- click on _Add New_
- fill the form. Using an image is optional, but it should open the media library if clicked.

The widget preview on the customizer should clear out, and display the entered values as they are typed on the form.

- click _Save_

the form should close, and the new SP button should be added to the drop down list. The customizer preview should show the new SP button.

- click _Cancel_

the form should clear and close, and the previously selected SP button should appear on the customizer preview.

To **edit and existing product**, you'll need to:

- Select the desired product from the drop down list
- click on _Edit Selected_

The form should load the product properties, and the widget preview on the customizer should show the correct product, and update the values as they are edited on the form.

- click _Save_

the form should close, preserving the changes on the customizer preview window and the selected item on the product drop down list.

- click _Cancel_

the form should clear and close, and the previously selected SP button should appear on the customizer preview.

To **delete an existing product**, you'll need to:

- Select the desired product from the drop down list.
- click on _Edit Selected_

The form should load the product properties, and the widget preview on the customizer should show the correct product.

- click _Delete_

After confirming the action, the selected product should disappear from the product drop down list. The first product on the list should be selected, and the customizer preview should reflect this change. 

#### Known Issues and pending items:

- [x] #9714 (Widgets: Make the Simple Payments button compatible with Selective Refresh)
- [x] If the selected Payment Button is deleted, the widget just shows the title and empty content.
- [x] https://github.com/Automattic/wp-calypso/issues/25327 (Customizer: Simple Payments Widget do not show options when the user has no existing buttons.)
- [x] https://github.com/Automattic/wp-calypso/issues/25484 (Stats For Simple Payments in Calypso)
- [ ] https://github.com/Automattic/jetpack/issues/9777 (Customizer: add WP-Admin/Admin user only warnings for Simple Payments)
- [x] https://github.com/Automattic/wp-calypso/issues/25576 (Customizer: Simple Payments Widget does not work on WP-Admin's Appearance > Widgets interface.)
#### Proposed changelog entry for your changes:

- Fixes an error when the user has published Simple Payment Products but the Professional Subscription has expired.
- Adds Simple Payment Products as Widgets, with the option to manage them via the Customizer.
